Bluetooth: Host: Fix issues with host IRK handling

joerchan opened the pull request:
Fix multiple issues related to the way the host handles
Identity Information related to privacy

1. If the controller provided a public address the IRK
   for this identity would be randomly generated but not
   stored persistenly.

2. Fix the handling of the above issue which was fixed
   for the random address but would initiate settings save
   ID on every boot.

3. Fix the host not using the Vendor Specific HCI commands
   related to retrieving the Identity Root (IR) from the
   controller and using the key diversified function d1
   to generate an IRK as specified in the BT Core spec.

The following behaviours currently exist, and I'm unsure if they have to be solved or can be accepted.

If the application creates IDs after enable. Then the identities created during bt_enable will also be stored, even though they were not randomly generated.

If the application creates IDs before enable but does not provide either Address or IRK then this part will be randomly generated. This ID will not be persisted and a new one will be generated on next reboot.

If the application creates IDs before enable and after enable, then all IDs created before enable will also be stored persistently.

joerchan committed:
Bluetooth: controller: Return FICR as a static address IR

Return the IR defined in FICR as the Identity Root for the static
address through the read static addresses command instead of providing
it through the Read Key Hierarchy Root command.
This is following the recommendations in the Zephyr HCI extension
document in doc/reference/bluetooth/hci.txt

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: Host: Fix issues with host IRK handling

Fix multiple issues related to the way the host handles
Identity Information related to privacy

1. If the controller provided a public address the IRK
   for this identity would be randomly generated but not
   stored persistenly.

2. Fix the handling of the above issue which was fixed
   for the random address but would initiate settings save
   ID on every boot.

3. Fix the host not using the Vendor Specific HCI commands
   related to retrieving the Identity Root (IR) from the
   controller and using the key diversified function d1
   to generate an IRK as specified in the BT Core spec.

Make sure that a Host generated ID is only saved when it is first
generated.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4686,6 +4659,36 @@ int bt_addr_le_create_static(bt_addr_le_t *addr)
 	return 0;
 }
 
+static u8_t bt_read_public_addr(bt_addr_le_t *addr)
```
Why u8_t? That seems like a strange choice. The usual `int`-style error reporting would seem appropriate here.

joerchan replied:
It matches bt_read_static_addr, returning the number of addresses read. there can be multiple static addresses, but only one public one. So in truth this could be a bool, I'm assigning bt_dev.id_count to the return value of this function.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4686,6 +4659,36 @@ int bt_addr_le_create_static(bt_addr_le_t *addr)
+
+	BT_DBG("status 0x%02x", rp->status);
+
+	if (rp->status ||
```
`bt_hci_cmd_send_sync()` will return an error if the command complete status was non-zero, so `rp->status` will never be non-zero at this point. (it's a separate discussion whether this is a good thing, since it makes it impossible to catch the exact HCI status later, the HCI command API only has two possible mappings for it, if you check the implementation)

joerchan replied:
Correct, thanks.

I included the Logging of the `BT_DBG("status 0x%02x", rp->status);` from bdaddr and mistankenly thought I had to handle this status.
Looking over the code there are more instances similar to `read_baddr_complete` that logs the HCI status code, but it is no point, because the return error code has to be handled and the buf is always unref. So we could remove all of those.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5464,13 +5445,80 @@ int bt_id_delete(u8_t id)
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_VS_READ_KEY_HIERARCHY_ROOTS, NULL,
+		&rsp);
```
The split line seems incorrectly aligned

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5464,13 +5445,80 @@ int bt_id_delete(u8_t id)
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_VS_READ_KEY_HIERARCHY_ROOTS, NULL,
+		&rsp);
+	if (err) {
+		return;
```
perhaps log also that we got an error?

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5464,13 +5445,80 @@ int bt_id_delete(u8_t id)
+	rp = (void *)rsp->data;
+
+	BT_DBG("status 0x%02x", rp->status);
+	if (rp->status) {
```
This is redundant since `rp->status` will always be zero if `bt_hci_send_cmd_sync()` returned zero

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5481,44 +5529,72 @@ static uint8_t bt_read_static_addr(bt_addr_le_t *addr)
 
+	BT_DBG("status 0x%02x", rp->status);
+
+	if (rp->status) {
```
Redundant check (see my other similar comments)

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -30,7 +30,7 @@ enum {
 	BT_DEV_READY,
 	BT_DEV_PRESET_ID,
-	BT_DEV_USER_ID_ADDR,
+
```
Any particular reason you're changing the grouping here by adding an empty line?

joerchan replied:
Didn't mean to do that, reverted

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -463,6 +463,25 @@ static int bt_smp_aes_cmac(const u8_t *key, const u8_t *in, size_t len,
+
+	sys_put_le16(d, &res[0]);
+	sys_put_le16(r, &res[2]);
+	memset(&res[4], 0, 16-4);
```
minor: `16 - 4`

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4786,16 +4789,18 @@ static void hci_vs_init(void)
+	    (bt_dev.hci_version < BT_HCI_VERSION_5_0)) {
+		bt_addr_le_t addr;
+
+		if (!bt_read_public_addr(&addr)) {
```
I'm not following the logic here. If the HCI version is less than 5.0 the controller will pretty certainly not support the Zephyr vendor extensions, so why bother sending the command? The vendor opcodes address space is freely definable per-vendor basis, so sending vendor commands can generate wildly unexpected results if the controller you're talking does support vendor commands but uses a completely different vendor command set. What makes things worse is that you're not checking the length of the returned parameters, so if this happens to map to a non-Zephyr command that returns a shorter set of parameters the code will read past the valid part of the buffer (in addition to this code falsely concluding that the Zephyr extensions are supported).

joerchan replied:
You are right. I did this wrong.

Your point about length checking is very valid. It is not something I changed in this PR. But since I'm including new VS HCI commands I found it easiest to include a length check commit here.

I added length check when CONFIG_BT_HCI_VS_EXT_DETECT is used on all commands. Maybe it is a bit excessive but I found it safer than only checking on the first command complete for info. Might have been a coincidence.

joerchan replied:
The heuristic should be that if the Version is 5.0 or later, and the controller does not have a valid public address. It's not a very good heuristic in my opinion. It works for this case because we only use to check for random static identity address. But in the case the controller supports many extensions to HCI but still has a valid Bluetooth address we would not be able to use these features.

jhedberg replied:
Yep, there really isn't any good solution here (if there was we wouldn't need a Kconfig to enable the heuristics). One thing we do in Linux is to check the reported manufacturer ID, however Zephyr allows overriding it so it's not particularly helpful.

jhedberg replied:
Hmm, maybe we should require the LF manufacturer ID by default for the heuristics? If someone wants to pair up Zephyr host with a controller with non-LF manufacturer ID then they won't be able to use the heuristics but just have the host assume that it's talking to a Zephyr?

jhedberg replied:
IIRC the Zephyr controller was introduced before LF had its SIG-assigned ID, which is why this code isn't taking advantage of it. Perhaps we can simply assume no-one is going to pair up a recent Zephyr host with an ancient Zephyr controller.

joerchan replied:
I think the LF company ID is too restrictive. It would prevent us from having other controllers implement the Zephyr HCI specified VS extension. In my opinion the best approach would be to have a kind of magic number. But that would mean adding a read_magic command/event.

I will create an issue to handle this outside of this PR and leave it like it already was.

joerchan committed:
Bluetooth: HCI: Check length of VS command complete

Always check that the length of the returned command complete
event for a vendor specific command matches the expected length
when the support for Zephyr VS HCI commands are uncertain.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
Sorry for the mess, I messed up a force-push by not including changes I had already made.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5464,13 +5466,81 @@ int bt_id_delete(u8_t id)
+	struct bt_hci_rp_vs_read_key_hierarchy_roots *rp;
+	struct net_buf *rsp;
+	int err;
+#endif /* defined(CONFIG_BT_HCI_VS_EXT) */
```
For the record, I know I've been trying to stick to the rule of having variable declarations in the beginning of a scope, however that does not seem to be a requirement for Zephyr. There's e.g. many places in the core kernel code where the variable declarations are closer to the point where they're used. If you moved these after the memset you'd only need to have a single ifdef section.

joerchan replied:
Thank you for the clarification. I have moved them to reduce the ifdefs

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5464,13 +5466,81 @@ int bt_id_delete(u8_t id)
+#if defined(CONFIG_BT_PRIVACY)
+	u8_t ir_irk[16];
+	u8_t ir[16];
+#endif
```
Same here regarding positioning of variable declarations and ifdefs

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4775,6 +4777,7 @@ static const char *vs_fw_variant(u8_t variant)
 }
 #endif /* CONFIG_BT_DEBUG */
 
+
```
stray extra line?

carlescufi reviewed (approved):
I only reviewed the changes to the HCI driver code.

carlescufi commented:
> I only reviewed the changes to the HCI driver code.

That was quite a oversight and misreading of the VS spec @joerchan, nice catch.

joerchan commented:
@jhedberg Did you check the question I posed in the description.
Specifically this one I'm interested in:
 - If the application creates IDs before enable but does not provide either Address or IRK then this part will be randomly generated. This ID will not be persisted and a new one will be generated on next reboot.

jhedberg commented:
> @jhedberg Did you check the question I posed in the description.

@joerchan no, I missed that part. The original intention has been that the only case where we don't store in flash is when the app provides existing values (hard-coded, or from some other storage), rather than the stack generating random ones. However, if this wasn't working before this PR then I don't think it's a reason to prevent it from being merged.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5481,44 +5547,81 @@ static uint8_t bt_read_static_addr(bt_addr_le_t *addr)
 
+	if (IS_ENABLED(CONFIG_BT_HCI_VS_EXT_DETECT) &&
+	    rsp->len != (sizeof(struct bt_hci_rp_vs_read_static_addrs) +
+			 cnt * sizeof(struct bt_hci_vs_static_addr))) {
```
This isn't quite right if `CONFIG_BT_ID_MAX` is less than `rp->num_addrs`. So what you really should be multiplying by here is `rp->num_addrs` and not `cnt`

joerchan replied:
Fixed.
